src/Aspire.Hosting.Azure/Provisioning/Internal/DefaultProvisioningContextProvider.cs

```diff
@@ -26,8 +31,157 @@ internal sealed class DefaultProvisioningContextProvider(
+                _provisioningOptionsAvailable.SetException(ex);
+            }
+        });
+    }
+
+    private async Task RetrieveAzureProvisioningOptions(JsonObject userSecrets, CancellationToken cancellationToken = default)
+    {
+        while (_options.Location == null || _options.SubscriptionId == null)
+        {
+            var locations = typeof(AzureLocation).GetProperties(BindingFlags.Public | BindingFlags.Static)
+                            .Where(p => p.PropertyType == typeof(AzureLocation))
+                            .Select(p => (AzureLocation)p.GetValue(null)!)
+                            .Select(location => KeyValuePair.Create(location.Name, location.DisplayName ?? location.Name))
+                            .OrderBy(kvp => kvp.Value)
+                            .ToList();
+
+            var messageBarResult = await interactionService.PromptMessageBarAsync(
+                 "Azure provisioning",
+                 "The model contains Azure resources that require an Azure Subscription.",
+                 new MessageBarInteractionOptions
+                 {
+                     Intent = MessageIntent.Warning,
+                     PrimaryButtonText = "Enter values"
+                 },
+                 cancellationToken)
+                 .ConfigureAwait(false);
+
+            if (messageBarResult.Canceled)
+            {
+                // User canceled the prompt, so we exit the loop
+                _provisioningOptionsAvailable.SetException(new MissingConfigurationException("Azure provisioning options were not provided."));
+                return;
+            }
+
+            if (messageBarResult.Data)
+            {
+                var result = await interactionService.PromptInputsAsync(
+                    "Azure provisioning",
+                    """
+                    The model contains Azure resources that require an Azure Subscription. 
+
+                    To learn more, see the [Azure provisioning docs](https://aka.ms/dotnet/aspire/azure/provisioning).
+                    """,
+                    [
+                        new InteractionInput { InputType = InputType.Choice, Label = "Location", Placeholder = "Select Location", Required = true, Options = [..locations] },
+                        new InteractionInput { InputType = InputType.SecretText, Label = "Subscription ID", Placeholder = "Select Subscription ID", Required = true },
+                        new InteractionInput { InputType = InputType.Text, Label = "Resource group", Value = GetDefaultResourceGroupName()},
+                    ],
+                    new InputsDialogInteractionOptions
+                    {
+                        EnableMessageMarkdown = true,
+                        ValidationCallback = static (validationContext) =>
+                        {
+                            var subscriptionInput = validationContext.Inputs[1];
```

Comment:
Instead of referencing inputs by index, you could create them earlier and assign them to variables, use the variable in the array passed to PromptInputsAsync and here in the validation

Although I'm starting to wonder if inputs should have a name and be accessible via name here and in the result...